### ISSUE TITLE
[FEATURE] discussion with dynamic efficiency improvement

### ISSUE THREAD

**Issue by weiyusheng** [MouseLand/cellpose#1435]
English is not my native language, please excuse my spell errors.

# problem

I found it slow with cpsam model.

So I run this model with:

```
with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA], record_shapes=True) as prof:
    with torch.inference_mode():
        with torch.autocast(device_type='cuda', dtype=torch.float16):
            masks, flows, styles = model.eval(...)
print(prof.key_averages().table(sort_by="cuda_time_total", row_limit=10))
```

env:

```
cpu i7-11700
gpu RTX 4070
gpu driver 595.97
cuda 13.2
2d (2k,2k) image
fp16 eval
```

And I find that plenty of time is cost during Aten::_copy both on CPU and GPU.

--------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                              Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls
                                       aten::copy_        26.98%        3.332s        27.06%        3.342s     146.784us        2.059s        14.12%        2.069s      90.865us         22766
                                         aten::add         2.93%     361.768ms         3.36%     414.731ms      18.883us        1.707s        11.71%        1.763s      80.263us         21963
                                      aten::select        11.11%        1.373s        11.40%        1.408s      16.451us        1.109s         7.60%        1.694s      19.796us         85569
                                          aten::mm         0.66%      82.099ms         0.66%      82.099ms      77.745us     894.875ms         6.14%     894.875ms     847.420us          1056
                                       aten::copy_        26.98%        3.332s        27.06%        3.342s     146.784us        2.059s        14.12%        2.069s      90.865us         22766
                                         aten::add         2.93%     361.768ms         3.36%     414.731ms      18.883us        1.707s        11.71%        1.763s      80.263us         21963
                                      aten::select        11.11%        1.373s        11.40%        1.408s      16.451us        1.109s         7.60%        1.694s      19.796us         85569
                                       aten::copy_        26.98%        3.332s        27.06%        3.342s     146.784us        2.059s        14.12%        2.069s      90.865us         22766
                                         aten::add         2.93%     361.768ms         3.36%     414.731ms      18.883us        1.707s        11.71%        1.763s      80.263us         21963
                                       aten::copy_        26.98%        3.332s        27.06%        3.342s     146.784us        2.059s        14.12%        2.069s      90.865us         22766
                                       aten::copy_        26.98%        3.332s        27.06%        3.342s     146.784us        2.059s        14.12%        2.069s      90.865us         22766
                                       aten::copy_        26.98%        3.332s        27.06%        3.342s     146.784us        2.059s        14.12%        2.069s      90.865us         22766
                                       aten::copy_        26.98%        3.332s        27.06%        3.342s     146.784us        2.059s        14.12%        2.069s      90.865us         22766
                                         aten::add         2.93%     361.768ms         3.36%     414.731ms      18.883us        1.707s        11.71%        1.763s      80.263us         21963
                                      aten::select        11.11%        1.373s        11.40%        1.408s      16.451us        1.109s         7.60%        1.694s      19.796us         85569
--------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------

With further tests, here are some performance bottleneck:

(by the way, there has been a long time from the last pr that changes the "dynamics.py" file, so the following discussion could be thought as done on the newest commit such as https://github.com/MouseLand/cellpose/commit/a9f8bfcde43033247309e3982747df9fe9f09315)

### cellpose/dynamics.py 355

```
    for t in range(niter):
        dPt = torch.nn.functional.grid_sample(im, pt, align_corners=False)
        for k in range(ndim):  #clamp the final pixel locations
            pt[..., k] = torch.clamp(pt[..., k] + dPt[:, k], -1., 1.)
```

Well, there is not an efficiency problem. But with a big niter, it must be slow.(niter=2000 in my test)

```
eps = 0.1
max_v = max(shape)
deps = eps / max_v
for t in range(niter):
    ***some code***
    if t % 20 == 0 and torch.max(torch.abs(dPtt)) < deps:
        break
```

<img width="1563" height="711" alt="Image" src="https://github.com/user-attachments/assets/18741132-9e72-45d4-9b52-49c3143c00a7" />

I think it useful to stop this iter earlier. After this function, the result will be tranformed into int, so some tiny residuals will not be a problem.(eps could be set higher)

In my test, this iter decreases from 2000 to nearly 400 without influence.

### cellpose/dynamics.py 543

```
    for k in range(n_seeds):
        slc = tuple([slice(seeds1[k][j]-5, seeds1[k][j]+6) for j in range(ndim)])
        h_slc[k] = h1[slc]
```

It runs slower on GPU(the whole get_masks_torch function tasks about 2s) than on CPU(about 0.2s) for maybe data transfer.

Here are two ways to solve this. One is add "pt = pt.cpu() at the beginning of the get_masks_torch function", another is:

```
if ndim == 2:
    sd_0_clp = torch.clamp(seeds1[:, 0], min = 5, max = shape[0]-6)
    sd_1_clp = torch.clamp(seeds1[:, 1], min = 5, max = shape[1]-6)
    offset_t = torch.arange(11, device=seeds1.device) - 5
    idx_d = sd_0_clp[:,None] + offset_t
    idx_h = sd_1_clp[:,None] + offset_t
    patchs = h1[idx_d[:,:,None], idx_h[:,None,:]]
if ndim == 3:
    sd_0_clp = torch.clamp(seeds1[:, 0], min = 5, max = shape[0]-6)
    sd_1_clp = torch.clamp(seeds1[:, 1], min = 5, max = shape[1]-6)
    sd_2_clp = torch.clamp(seeds1[:, 2], min = 5, max = shape[2]-6)
    offset_t = torch.arange(11, device=seeds1.device) - 5
    idx_d = sd_0_clp[:,None] + offset_t
    idx_h = sd_1_clp[:,None] + offset_t
    idx_w = sd_2_clp[:,None] + offset_t
    patchs = h1[idx_d[:,:,None,None], idx_h[:,None,:,None], idx_w[:,None,None,:]]
```

<img width="1500" height="818" alt="Image" src="https://github.com/user-attachments/assets/1a9f304c-dfeb-4f89-badb-f50738d6aa20" />

### cellpose/dynamics.py 558

```
    seeds_new = [tuple((torch.nonzero(seed_masks[k]) + seeds1[k] - 5).T) 
            for k in range(n_seeds)]
    del seed_masks 
    
    dtype = torch.int32 if n_seeds < 2**16 else torch.int64
    M1 = torch.zeros(shape, dtype=dtype, device=device)
    for k in range(n_seeds):
        M1[seeds_new[k]] = 1 + k
```

The same as above, running slow on GPU, especially the first line.

```
dtype = torch.int32 if n_seeds < 2**16 else torch.int64
M_t = torch.zeros(shape, dtype=dtype, device=device)
test = torch.nonzero(seed_masks).to(dtype)

k_idx = test[:, 0]
yx = test[:, 1:]

offset = seeds1[k_idx]
idx_t = yx + offset - 5

values = k_idx + 1

H, W = M_t.shape

linear_idx = idx_t[:,0] * W + idx_t[:,1]

values = k_idx + 1

M_flat = torch.zeros(H * W, device=M_t.device, dtype=values.dtype)

M_flat.scatter_reduce_(
    0,
    linear_idx,
    values,
    reduce="amax",
    include_self=False
)

M_t = M_flat.view(H, W)
```
((H,W)only for 2d, need (D,H,W) for 3d)

<img width="1562" height="970" alt="Image" src="https://github.com/user-attachments/assets/84ab7020-30d7-44cb-af38-5bf80c1343dc" />

<img width="1124" height="722" alt="Image" src="https://github.com/user-attachments/assets/72e964e6-1316-4913-8e6e-f505b47b3dd6" />

### cellpose/dynamics.py 42

```
    for i in range(n_iter):
        T[tuple(meds.T)] += 1
        Tneigh = T[tuple(neighbors)]
        Tneigh *= isneighbor
        T[tuple(neighbors[:, 0])] = Tneigh.mean(axis=0)
```

It's slow but I have not an optimization for it.

I'm trying to optimize the masks_to_flows_gpu function, aiming to bypass neighbors and isneighber.

If anyone has an idea related, I'd like to learn about it.

**Comment by carsen-stringer** [comment]
Thanks @weiyusheng , there were definitely some inefficient things in dynamics.py that I was doing - the biggest one was in extend_centers_gpu with the tuple indexing - claude suggested using flattened indexing there and that helped throughout, including the seeds. Happy to get more feedback on how to speed things up, for now I'm going to close this issue.

Btw the number of iterations is up to the user - we have a default of 200 but you can change that

**Comment by weiyusheng** [comment]
Thanks for your kind response.

It's pity that I don't get a novel speed up with the new dynamics.py. It takes a little bit more time in fact, and here is the result:

before:

<img width="1523" height="286" alt="Image" src="https://github.com/user-attachments/assets/c137ab62-d5fa-4b03-b631-29e3b10df6a0" />

after:

<img width="1464" height="286" alt="Image" src="https://github.com/user-attachments/assets/459445a7-1af2-41c8-a6f1-98088e9d9465" />

Thanks again.

**Comment by carsen-stringer** [comment]
Maybe it's a different in GPUs, but with a 5000x5000 image containing ~4000 masks, compute_masks now takes ~1.1 sec rather than ~1.8 sec (with flow_threshold=0.4). With a 20000x20000 image containing ~70000 masks, it is ~13 sec rather than 27 sec. We think this is fast enough for most users, especially given running the network is the majority of the runtime.
